src/Features/Core/Portable/CodeRefactorings/AbstractRefactoringHelpersService.cs

```diff
@@ -25,39 +23,27 @@ internal abstract class AbstractRefactoringHelpersService<TExpressionSyntax, TAr
 
     public abstract bool IsBetweenTypeMembers(SourceText sourceText, SyntaxNode root, int position, [NotNullWhen(true)] out SyntaxNode? typeDeclaration);
 
-    public async Task<ImmutableArray<TSyntaxNode>> GetRelevantNodesAsync<TSyntaxNode>(
-        Document document, TextSpan selectionRaw, bool allowEmptyNodes, CancellationToken cancellationToken) where TSyntaxNode : SyntaxNode
+    private static void AddNode<TSyntaxNode>(bool allowEmptyNodes, ref TemporaryArray<TSyntaxNode> result, TSyntaxNode node) where TSyntaxNode : SyntaxNode
     {
-        using var _1 = ArrayBuilder<TSyntaxNode>.GetInstance(out var relevantNodesBuilder);
-        await AddRelevantNodesAsync(document, selectionRaw, relevantNodesBuilder, cancellationToken).ConfigureAwait(false);
-
-        if (allowEmptyNodes)
-            return relevantNodesBuilder.ToImmutableAndClear();
-
-        using var _2 = ArrayBuilder<TSyntaxNode>.GetInstance(out var nonEmptyNodes);
-        foreach (var node in relevantNodesBuilder)
-        {
-            if (node.Span.Length > 0)
-                nonEmptyNodes.Add(node);
-        }
+        if (!allowEmptyNodes && node.Span.IsEmpty)
+            return;
 
-        nonEmptyNodes.RemoveDuplicates();
-        return nonEmptyNodes.ToImmutableAndClear();
+        result.Add(node);
     }
 
-    private async Task AddRelevantNodesAsync<TSyntaxNode>(
-        Document document, TextSpan selectionRaw, ArrayBuilder<TSyntaxNode> relevantNodes, CancellationToken cancellationToken) where TSyntaxNode : SyntaxNode
+    public void AddRelevantNodes<TSyntaxNode>(
+        ParsedDocument document, TextSpan selectionRaw, bool allowEmptyNodes, bool stopOnFirst, ref TemporaryArray<TSyntaxNode> result, CancellationToken cancellationToken) where TSyntaxNode : SyntaxNode
```

Comment:
passing in PArsedDocument allows this to all be synchronous.  which also allows passing in TempArray to add results into.